actions/src/sema4ai/actions/agent/__init__.py

```diff
@@ -0,0 +1,234 @@
+        )
+
+    # If we have an invocation context, we should always have the id!
+    id_value = invocation_context.get(id_key)
+
+    if not id_value:
+        raise ActionError(
+            f"Internal Error: No {id_key} found in the invocation_context!"
+        )
+
+    if not isinstance(id_value, str):
+        raise ActionError(
+            f"Internal Error: {id_key} is not a string (found: {id_value} ({type(id_value)}))!"
+        )
+    return id_value
+
+
+def _get_id_generic(id_key: str, header_key: str) -> str:
+    from sema4ai.actions._action import get_current_requests_contexts
+
+    id_value = None
+    request_contexts = get_current_requests_contexts()
+    if request_contexts is None:
+        raise ActionError(
+            f"Internal Error: Unable to get the {id_key} (no context available)"
+        )
+
+    # Prefer invocation context.
+    if request_contexts.invocation_context:
+        id_value = _get_id_from_invocation_context(
+            request_contexts.invocation_context, id_key
+        )
+
+    if id_value is None:
+        # Alternative use case when coupled directly with the
+        # agent server without a router to make the invocation/action context
+        # available.
+        request = request_contexts._request
+        if request:
+            id_value = request.headers.get(header_key)
+
+    if id_value is None:
+        action_context = request_contexts.action_context
+        if action_context:
+            # Old use case (without separate id in the action context)
+            id_value = _get_id_from_action_context(action_context, id_key)
+
+    if id_value is None:
+        # Ok, unable to get the id through any of the heuristics,
+        # let's raise an error.
+        raise ActionError(
+            f"Internal Error: Unable to get the {id_key} from the action context or the request headers"
+        )
+
+    return id_value
+
+
+def get_thread_id() -> str:
+    """Get the thread ID from the action context or the request headers.
+
+    Note:
+        This will raise an ActionError if the thread ID cannot be found. This is expected when calling this
+        function directly from VSCode unless the `x-invoked_for_thread_id` header is set in the request in configured
+        inputs.
+    """
+    return _get_id_generic("thread_id", "x-invoked_for_thread_id")
+
+
+def get_agent_id() -> str:
+    """Get the agent ID from the action context or the request headers.
+
+    Note:
+        This will raise an ActionError if the agent ID cannot be found. This is expected when calling this
+        function directly from VSCode unless the `x-invoked_by_assistant_id` header is set in the request in configured
+        inputs.
+    """
+    return _get_id_generic("agent_id", "x-invoked_by_assistant_id")
+
+
+def prompt_generate(
+    prompt: Prompt | dict,
+    platform_config: AnyPlatformParameters | dict | None = None,
+    thread_id: str | None = None,
+    agent_id: str | None = None,
+) -> dict:
```

Comment:
Can we get the actual shape of the return in typing as a proper typeddict instead of a plain `dict`?